### ISSUE TITLE
Make Publii Progressive Web App Compatible

### ISSUE THREAD

**Issue by JOduMonT** [GetPublii/Publii#340]
Would be awesome if Publii was PWA *Progressive Web App* compatible.

1. this is help the SEO (because Google focus on mobile 1st)
2. it speed up the loading time (because of the caching)
3. it's the new mobile first rule since last year ;)

To do so it is more or less following this [checklist](https://developers.google.com/web/progressive-web-apps/checklist)

#### Good News
Publii starts already with 54% done
it almost just missing a workers and a manifest.

![image](https://user-images.githubusercontent.com/5204724/53028828-ec2a9400-3467-11e9-9021-528e937693a8.png)

**Comment by dziudek** [comment]
PWA is as the name suggest - for Apps. I do not see any sense for creating PWA manifest for normal websites. Google looks for the mobile speed index and Publii themes (if well managed) gets amazing results in that test (usually over 90/100). 

PWA for normal blog is definitely "just for fun" solution. The only profit from it is offline access, but to be honest - how often people are offline today and even if they are - somebody really believe that people who are offline, will really want to visit their blog in that offline time?

PWA in Publii will have sense if Publii will support app-like themes - it is a future at this moment.

**Comment by jodumont** [comment]
thanks for your prompt feedback

**Comment by dzmitry-lahoda** [comment]
Will PWA give me changec to distribute site as Android Application? Would be fun. And if that app has local search #321 . I guess I can create custom html to invoke offline OSM maps application:)

**Comment by jodumont** [comment]
Yes; a site "PWA compatible" is able to works offline and to put a shortcut on Desktop / Android and probably iPhone too.

I you want to try just go at this [website](https://pwa-jekyll-starter.netlify.com/) with Chrome then you will be able to "Install the Site on your computer" and consulting it offline

![image](https://user-images.githubusercontent.com/5204724/53285090-b97ce600-375c-11e9-8ceb-96d9a8997a2f.png)

#### Why I'm asking it

Know I know it's very specific to my case but; I'm doing howto for sysadmin and devops which they don't always have to reflex to save the site and/or print it as PDF before starting it. And When it comes to mess with firewall, internet connection and/or your in a server room; your not always able to relay on Internet; so this is why I tough of this PWA feature. But Hexo do the job well so, I'll probably stick with it for now.

**Comment by YoavRheims** [comment]
Putting my two cents, [Google's Workbox](https://developers.google.com/web/tools/workbox/) is a pretty neat solution to begin with PWA. It's quite simple to implement.

**Comment by inY2K** [comment]
I saw on others mediums that the dev is really against this idea; however, I urged him to consider this one more time. There are many cases where PWA is a fantastic idea. And yes, people would love to be able to open and read their favorite blogs. Many static web sites generator like Bootstrap studio are implementing PWA. It's fun, and it makes sense. Blazon from Microsoft has a similar function, as well. The other good reason to achieve this is, why not? Just because one isn't inclined to a particular technology, doesn't mean that it doesn't make sense.  I love Publii and would use it over any other system, but PWA will be fantastic to have!

**Comment by antoninadert** [comment]
I add my 2 cents since I see a use case for PWA on the simple theme: a photo galery app.
It is so cool to showcase your work with an app directly.
So I believe proposing PWA would be a nice addition.
Not necessarily a must have, but definitely a nice to have, especially for mobile users

**Comment by JOduMonT** [comment]
### PWA is the new kids in the block for SEO such as was OpenGraph, Schema Markup and AMP few years ago.

So it's not only related to being offline and/or make a app

For me it was Publii claim good with SEO and follow the latest implementation related to SEO.
![image](https://user-images.githubusercontent.com/5204724/75110837-a9e0f780-5665-11ea-8455-c7bb5ece24f4.png)

**Comment by dziudek** [comment]
@JOduMonT - please prove this words "PWA is the new kid in the block for SEO such as was OpenGraph, Schema Markup and AMP few years ago." - and I mean strong and proven evidence to support your thesis, not some suspicious and vague theories.

Also since PWAs looks like typical shortcuts on newer Android devices it makes them even less attractive than earlier to implement: https://android.stackexchange.com/questions/209408/how-to-create-app-like-chrome-shortcuts

**Comment by JOduMonT** [comment]
@dziudek > please prove this words
actually your are right, PWA it self don't improve SEO if your website comply with PWA recommendation this one will be only faster more secure and user friendly which are 3 main factors on how google rate website.

1. **Fast**er: PWAs will respond quickly to user interactions, independently of network speed, location or device
2. More **Reliable**: when launched from the user’s home screen, service workers enable the PWA to load instantly, regardless of the network state (more on this later)
3. Installable: they have the ability to live on the user’s home screen, without the need for going through the app store
4. More **Immersive** and Engaging: they offer an immersive full-screen experience and can re-engage users with web push notifications
5. Progressive: they **work for every** user, regardless of the **browser**
6. **Responsive**: they work across desktop, mobile and tablet
7. Discoverable: their content can be findable through search
8. Linkable: it’s easy to share the application via URL
9. App-like: they look like a native app, thanks to the app shell model, which separates the app’s functionality from its content
10. Fresh: when new content is published and the user is connected to the Internet, that content will be made available in the app
11. Safe: because network requests can be intercepted through service workers, it’s a critical requirement that the app is served via HTTPS (no mixed content)

**Comment by dziudek** [comment]
@JOduMonT 

1. Google AFAIK during Page Speed test does not measure cache and any other PWA-powered features

2. as above

3. Do you know people who use this feature? People does not install normal apps and you believe that they will install websites on home screen? Especially for small blog? https://techcrunch.com/2017/08/25/majority-of-u-s-consumers-still-download-zero-apps-per-month-says-comscore/

4. as above - for most websites typical user do not care about full screen and push notifications are the most irritating website feature ever: https://info.localytics.com/blog/the-inside-view-how-consumers-really-feel-about-push-notifications

5. Websites are also working for every user ;)

6. as above

7. as above - PWA does not add any additional improvement in this area

8. website can be also shared via URL

9. Nope. They look like the website designer designed them

10. False. It depends from the website architecture. Adding a PWA manifest and basic web workers do not add such features

11. On properly configured website you are always safe, also browsers are more and more restrict for the mixed content: https://www.searchenginejournal.com/google-block-mixed-content-warning/329055/

I do hope that now you understand why I do not see any real reason to provide PWA feature as a core Publii feature? :)

To be honest - I understand that for some people PWA feature can be important, but the only way it will appear in Publii will be a separate plugin ;)

**Comment by JOduMonT** [comment]
> To be honest - I understand that for some people PWA feature can be important, but the only way it will appear in Publii will be a separate plugin ;)

To be honest, this is fair enough for me
but to be really honest I drop using Publii the day I bought a theme.
